### ISSUE TITLE
LINUX/OSX : I don't hear any sound

### ISSUE THREAD

**Issue by probonopd** [ttdm/Symphonium#3]
Running Symphonium on Ubuntu 18.04.2 LTS (Bionic Beaver). It does recognize my MIDI controller and it does open a MIDI file, but then I click on play or play notes on the MIDI keyboard I don't hear any music coming out of my computer's speakers. How can I debug this?

**Comment by ttdm** [comment]
1/ The answer is probalby yes but did you configure your midi input and outputs using the "option->MIDI in.out" menu ?

2/ There is an option which is forward input sound to output which is not active by default. It should be one of the final options of the option bar, next to the options to handle intervals.  (i usually play with my headphones on my MIDI keyboard and if this option is ON, you hear yourself twice)

3/ The soft hasnt been tested on any linux distrib as stated in the readme. I saw that in your pull request you added the corresponding ALSA flag but there can still be some issue. I was doing it this afternoon so I will try it at home and fix if needed in the coming hours ;)

**Comment by ttdm** [comment]
I was thinking : 
RTMidi creates an object whether or not it correctly detect ALSA (and or the chosen MIDI controller), do you see the key you press in your MIDI keyboard in violet on your screen ? 

If yeah it means all libs are properly working and its a sound issue. If no, it probably means that RTMidi is constructing an empty object (which means that it isnt properly linked with ALSA).

**Comment by ttdm** [comment]
HI, 

As stated in the readme, it hasnt been tested before on any linux distrib. 

I just tested it and It looks like the only way to have an audio output at the moment (linux only; it's working on windows), is to select your keyboard as input and output (and then either using the built in speakers of your keyboards if you have some or connecting your headphone to your keyboard).  Doing it, you will have all the sounds (soft generated and played by yourself). 

I'm trying to fix this atm.

**Comment by probonopd** [comment]
Well I just have a MIDI controller connected to the PC via USB, it has no speakers built in. So I need the computer to produce all of the sound.

**Comment by ttdm** [comment]
So : 
The output of the application is a MIDI stream. In windows, when selecting their MIDI outport, it automatically synthesize and stream the resulting audio wave. 

In linux, neither ALSA or JACK do this. Instead, the MIDI stream should be forwarded to a synthesizer configured as a deamon such as timidity++. 
This is how to setup timidity++ for arch users : https://wiki.archlinux.org/index.php/Timidity 

I must say that there is a **noticeable delay** (0.2s at least!) by using timidity behind symphonium. **I advise to use the application listening the sound back from the keyboard if possible !**

I will need to include a novel 3rd-party to directly generate sound from MIDI. It will probably take some time.

**Comment by probonopd** [comment]
Thank you for the explanation. As I guess this software is geared toward newcomers, not professional audio people, I think that adding a simple built-in piano sound generator would increase the "plug and play" factor of this project tremendously.

**Comment by ttdm** [comment]
I totally agree, I expected it to work that way; as it is on windows which should account for at least 90% of the users. 

I also expected most people using it by using their MIDI keyboard speakers or plugging their headphones on it and not on their computer. 

For people like you plugging your headphones on your computer, i will add an option to balance the keyboard sounds and the midi track sound which you can't do at the moment and which is crucial. 

I just want to add that to reduce the delay, the timidity deamon should simply be used with the following parameters : -B2,8 -0s -iAD  (and optionally --sequencer-ports=1)
  which are the default debian parameters. so :

`timidity -B2,8 -0s -iAD --sequencer-ports=1`

Thanks for your feedbacks, i will see what i can do to add a sound generator ;)

**Comment by probonopd** [comment]
Thank you :+1:

**Comment by Thonnard** [comment]
I have the same issue (no sound) in Ubuntu Studio 21. For now, I use Qsynth to get sounds, which works pretty well (no delays etc.).
